src/rez/utils/pip.py

```diff
@@ -0,0 +1,705 @@
+
+
+def get_rez_requirements(installed_dist, python_version, name_casings=None):
+    """Get requirements of the given dist, in rez-compatible format.
+
+    Example result:
+
+        {
+            "requires": ["foo-1.2+<2"],
+            "variant_requires": ["future", "python-2.7"],
+            "metadata": {
+                # metadata pertinent to rez
+                ...
+            }
+        }
+
+    Each requirement has had its package name converted to the rez equivalent.
+    The 'variant_requires' key contains requirements specific to the current
+    variant.
+
+    TODO: Currently there is no way to reflect extras that may have been chosen
+    for this pip package. We need to wait for rez "package features" before this
+    will be possible. You probably shouldn't use extras presently.
+
+    Args:
+        installed_dist (`distlib.database.InstalledDistribution`): Distribution
+            to convert.
+        python_version (`Version`): Python version used to perform the
+            installation.
+        name_casings (list of str): A list of pip package names in their correct
+            casings (eg, 'Foo' rather than 'foo'). Any requirement whose name
+            case-insensitive-matches a name in this list, is set to that name.
+            This is needed because pip package names are case insensitive, but
+            rez is case-sensitive. So a package may list a requirement for package
+            'foo', when in fact the package that pip has downloaded is called 'Foo'.
+            Be sure to provide names in PIP format, not REZ format (the pip package
+            'foo-bah' will be converted to 'foo_bah' in rez).
+
+    Returns:
+        Dict: See example above.
+    """
+    _system = System()
+    result_requires = []
+    result_variant_requires = []
+
+    # create cased names lookup
+    name_mapping = dict((x.lower(), x) for x in (name_casings or []))
+
+    # requirements such as platform, arch, os, and python
+    sys_requires = set(["python"])
+
+    # assume package is platform- and arch- specific if it isn't pure python
+    is_pure_python = is_pure_python_package(installed_dist)
+    if not is_pure_python:
+        sys_requires.update(["platform", "arch"])
+
+    # Note: This is supposed to give a requirements list that has already been
+    # filtered down based on the extras requested at install time, and on any
+    # environment markers present. However, this is not working in distlib. The
+    # package gets assigned a LegacyMetadata metadata object, and in that code
+    # path, this filtering doesn't happen.
+    #
+    # See: vendor/distlib/metadata.py#line-892
+    #
+    requires = installed_dist.run_requires
+
+    # filter requirements
+    for req_ in requires:
+        reqs = normalize_requirement(req_)
+
+        for req in reqs:
+            # skip if env marker is present and doesn't evaluate
+            if req.marker and not req.marker.evaluate():
```

Comment:
I have to admit that I'm confused here. Tell me if I have understood everything:

At this point, the extras have been removed from the marker. So the marker should only contain a python version and a platform?

Reply:
The marker can refer to a number of things - platform, python, os (see https://www.python.org/dev/peps/pep-0508/#id24). I'm not sure what you are getting at though? The marker might specify python-3, and we might be running python-2.7, in which case the req should be skipped.

In any case, I think I see a problem - the marker is going to get evaluated wrt current python interpreter, even though a rezified pip may have been used (and thus, a potentially different python version). We do avoid a lot of potential problems if the system pip (ie the one rez gets installed with) is only ever used, but then you're restricted to rez-pipping only for that version of python.

Reply:
We could override the python version all the time with the one we are running under or the one specified when `rez-pip` was invoked. So basically always do:

```python
req.marker.evaluate(
    environment={
        "python_full_version": str(python_version),
        "python_version": str(python_version.trim(2)),
        "implementation_version": str(python_version)
    }
)
```

Reply:
Yeah you read my mind, was hoping it was possible to do this way and looks like it is, cheers.